pkg/reconciler/route/reconcile_resources.go

```diff
@@ -106,34 +108,59 @@ func (c *Reconciler) reconcileIngress(
 	return ingress, effectiveRO, err
 }
 
-func (c *Reconciler) deleteServices(ctx context.Context, namespace string, serviceNames sets.String) error {
-	for _, serviceName := range serviceNames.List() {
-		if err := c.kubeclient.CoreV1().Services(namespace).Delete(ctx, serviceName, metav1.DeleteOptions{}); err != nil {
+func (c *Reconciler) deleteOrphanedServices(ctx context.Context, r *v1.Route, activeServices []resources.ServicePair) error {
+	ns := r.Namespace
+
+	active := make(sets.String, len(activeServices))
+
+	for _, service := range activeServices {
+		active.Insert(service.Service.Name)
+	}
+
+	routeLabelSelector := labels.SelectorFromSet(labels.Set{serving.RouteLabelKey: r.Name})
+	allServices, err := c.serviceLister.Services(ns).List(routeLabelSelector)
+
+	if err != nil {
+		return fmt.Errorf("failed to fetch existing services: %w", err)
+	}
+
+	for _, service := range allServices {
+		if active.Has(service.Name) {
+			continue
+		}
+
+		if err := c.kubeclient.CoreV1().Services(ns).Delete(ctx, service.Name, metav1.DeleteOptions{}); err != nil {
+			return fmt.Errorf("failed to delete Service: %w", err)
+		}
+
+		// Delete the endpoint if it exists
+		_, err := c.endpointsLister.Endpoints(ns).Get(service.Name)
+		if apierrs.IsNotFound(err) {
+			continue
+		} else if err != nil {
+			return fmt.Errorf("failed to list endpoint: %w", err)
+		}
+
+		if err := c.kubeclient.CoreV1().Endpoints(ns).Delete(ctx, service.Name, metav1.DeleteOptions{}); err != nil {
 			return fmt.Errorf("failed to delete Service: %w", err)
 		}
 	}
 
 	return nil
 }
 
-func (c *Reconciler) reconcilePlaceholderServices(ctx context.Context, route *v1.Route, targets map[string]traffic.RevisionTargets) ([]*corev1.Service, error) {
+func (c *Reconciler) reconcilePlaceholderServices(ctx context.Context, route *v1.Route, targets map[string]traffic.RevisionTargets) ([]resources.ServicePair, error) {
 	logger := logging.FromContext(ctx)
 	recorder := controller.GetEventRecorder(ctx)
+	ns := route.Namespace
 
-	existingServiceNames, err := c.getPlaceholderServiceNames(route)
-	if err != nil {
-		return nil, fmt.Errorf("failed to fetch existing services: %w", err)
-	}
+	services := make([]resources.ServicePair, 0, len(targets))
```

Comment:
not very important, but any reason to move this up here rather than leaving it about the loop that initialises it (rather than the one that initialises `names`)?

Reply:
Just grouping all the variable assignments that have the same scope - I'll tweak the whitespace in this function